docs/source/byo/interaction.rst

```diff
@@ -0,0 +1,228 @@
+            return -(h_proj - t_proj).squeeze(dim=-1).norm(p=2, dim=-1)
+
+Interactions with Different Dimension Vectors
+---------------------------------------------
+TransD is an example of an interaction module that not only uses two different
+representations for each entity and two representations for each relation, but they are
+of different dimensions.
+
+It can be implemented by choosing a different letter for use in the ``entity_shape`` and/or
+``relation_shape`` dictionary. Ultimately, the letters used are arbitrary, but you need to
+remember what they are when using the :func:`pykeen.models.make_model`,
+:func:`pykeen.models.make_model_cls`, or :func:`pykeen.pipeline.interaction_pipeline` functions
+to instantiate a model, make a model class, or run the pipeline using your custom interaction
+module (respectively).
+
+.. code-block:: python
+
+    from pykeen.nn.modules import Interaction
+    from pykeen.utils import project_entity
+
+    class TransDInteraction(Interaction):
+        entity_shape = ("d", "d")
+        relation_shape = ("e", "e")
+
+        def forward(self, h, r, t):
+            h, h_proj = h
+            r, r_proj = r
+            t, t_proj = t
+            h_bot = project_entity(
+                e=h,
+                e_p=h_p,
+                r_p=r_p,
+            )
+            t_bot = project_entity(
+                e=t,
+                e_p=t_p,
+                r_p=r_p,
+            )
+            return -(h_bot + r - t_bot).norm(p=2, dim=-1)
+
+.. note::
+
+    The :func:`pykeen.utils.project_entity` function was used in this implementation to
+    reduce the complexity. So far, it's the case that all of the models using multiple different representation
+    dimensions are quite complicated and don't fall into the paradigm of presenting simple examples.
+
+.. seealso:: A reference implementation is provided in :class:`pykeen.nn.modules.TransDInteraction`
+
+Differences between :class:`pykeen.nn.modules.Interaction` and :class:`pykeen.models.Model`
+-------------------------------------------------------------------------------------------
+The high-level :func:`pipeline` function allows you to pass pre-defined subclasses
+of :class:`pykeen.models.Model` such as :class:`pykeen.models.TransE` or
+:class:`pykeen.models.DistMult`. These classes are high-level wrappers around the interaction
+functions :class:`pykeen.nn.modules.TransEInteraction` and :class:`nn.modules.DistMultInteraction`
+that are more suited for running benchmarking experiments or practical applications of knowledge
+graph embeddings that include lots of information about default hyper-parameters, recommended
+hyper-parameter optimization strategies, and more complex applications of regularization schemas.
+
+As a researcher, the :class:`pykeen.nn.modules.Interaction` is a way to quickly translate
+ideas into new models that can be used without all of the overhead of defining a
+:class:`pykeen.models.Model`.
```

Comment:
Add
* Interaction functions can be used as (standalone) components outside of pykeen.

Reply:
93baa026